src/v/io/cache.h

```diff
@@ -0,0 +1,662 @@
+      boost::intrusive::constant_time_size<false>>;
+
+    bool evict_small() noexcept;
+    bool evict_main() noexcept;
+
+    size_t max_cache_size_;
+    size_t max_small_queue_size_;
+    size_t max_main_queue_size_;
+
+    Evictor evict_;
+    Cost cost_;
+
+    size_t small_queue_size_{0};
+    size_t main_queue_size_{0};
+    size_t ghost_queue_age_{0};
+
+    list_type small_fifo_;
+    list_type main_fifo_;
+};
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+cache<T, Hook, Evictor, Cost>::cache(
+  config config, const Evictor& evict, const Cost& cost) noexcept
+  : max_cache_size_(config.capacity)
+  , max_small_queue_size_(config.small_size)
+  , max_main_queue_size_(max_cache_size_ - max_small_queue_size_)
+  , evict_(evict)
+  , cost_(cost) {}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+void cache<T, Hook, Evictor, Cost>::insert(T& entry) noexcept {
+    while ((small_queue_size_ + main_queue_size_) > max_cache_size_) {
+        if (!evict()) {
+            break;
+        }
+    }
+
+    auto& hook = entry.*Hook;
+    if (on_ghost_queue(entry)) {
+        // evict from ghost queue
+        hook.insertion_time_.reset();
+
+        main_fifo_.push_back(entry);
+        hook.location_ = cache_hook::location::main;
+        main_queue_size_ += cost_(entry);
+    } else {
+        small_fifo_.push_back(entry);
+        hook.location_ = cache_hook::location::small;
+        small_queue_size_ += cost_(entry);
+    }
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+void cache<T, Hook, Evictor, Cost>::remove(const T& entry) noexcept {
+    const auto& hook = entry.*Hook;
+
+    if (hook.evicted()) {
+        return;
+    }
+
+    switch (hook.location_) {
+    case cache_hook::location::main:
+        main_queue_size_ -= cost_(entry);
+        main_fifo_.erase(main_fifo_.iterator_to(entry));
+        break;
+
+    case cache_hook::location::small:
+        small_queue_size_ -= cost_(entry);
+        small_fifo_.erase(small_fifo_.iterator_to(entry));
+        break;
+    }
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+bool cache<T, Hook, Evictor, Cost>::evict() noexcept {
+    if (small_queue_size_ > max_small_queue_size_) {
+        return evict_small();
+    }
+    return evict_main();
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+struct cache<T, Hook, Evictor, Cost>::stat
+cache<T, Hook, Evictor, Cost>::stat() const noexcept {
+    return {
+      .small_queue_size = small_queue_size_,
+      .main_queue_size = main_queue_size_,
+    };
+}
+
+template<
+  typename T,
+  cache_hook T::*Hook,
+  cache_evictor<T> Evictor,
+  cache_cost<T> Cost>
+bool cache<T, Hook, Evictor, Cost>::on_ghost_queue(
```

Comment:
nit: ghost_queue_contains, `on_ghost_queue` reads like some sort of callback

Reply:
Nice idea, I've changed this. I didn't like `on_ghost_queue` either for the same reason you mention.